envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/MetadataNodeGroup.kt

```diff
@@ -142,6 +142,10 @@ class MetadataNodeGroup(val properties: SnapshotProperties) : NodeGroup<Group> {
     private fun serviceName(metadata: NodeMetadata): String {
         return when (properties.incomingPermissions.enabled) {
             true -> metadata.serviceName.orEmpty()
+            // TODO: it is dangerous. Nobody remembers that serviceName is empty when incomingPermissions is disabled.
+            // Sooner or later somebody will use it to other purposes that incomingPermissions. We should validate
+            // that serviceName is always present (that means: envoys are required to provide serviceName)
+            // or signalise that it may be not set by changing type to String?
```

Comment:
Let me know what you think about this, which option do you prefer.

Reply:
Yes, but we can:
1. Validate it in NodeMetadataValidator and reject envoys without serviceName
2. Accept that it may be empty, and show it explicite by changing type to `String?`

If somebody wants to use EC without incoming-permissions enabled, not having a serviceName has a positive performance impact, because cache groups may be shared between different services. But I am not sure if we should focus on such use-cases.

Reply:
I created a task with solution (1): https://github.com/allegro/envoy-control/issues/91 and link it here in TODO comment.
If somebody has a different opinion we can still change it.